src/Features/LanguageServer/Protocol/Handler/Diagnostics/DiagnosticSources/AbstractWorkspaceDocumentDiagnosticSource.cs

```diff
@@ -40,14 +52,35 @@ public override bool IsLiveSource()
             }
             else
             {
-                // We call GetDiagnosticsForIdsAsync as we want to ensure we get the full set of diagnostics for this document
-                // including those reported as a compilation end diagnostic.  These are not included in document pull (uses GetDiagnosticsForSpan) due to cost.
-                // However we can include them as a part of workspace pull when FSA is on.
-                var documentDiagnostics = await diagnosticAnalyzerService.GetDiagnosticsForIdsAsync(
-                    Document.Project.Solution, Document.Project.Id, Document.Id,
-                    diagnosticIds: null, shouldIncludeAnalyzer, includeSuppressedDiagnostics: false,
-                    includeLocalDocumentDiagnostics: true, includeNonLocalDocumentDiagnostics: true, cancellationToken).ConfigureAwait(false);
-                return documentDiagnostics;
+                var projectDiagnostics = await GetProjectDiagnosticsAsync(diagnosticAnalyzerService, cancellationToken).ConfigureAwait(false);
+                return projectDiagnostics.WhereAsArray(d => d.DocumentId == Document.Id);
+            }
+        }
+
+        private async ValueTask<ImmutableArray<DiagnosticData>> GetProjectDiagnosticsAsync(
+            IDiagnosticAnalyzerService diagnosticAnalyzerService, CancellationToken cancellationToken)
+        {
+            if (!s_projectToDiagnostics.TryGetValue(Document.Project, out var lazyDiagnostics))
+            {
+                // Extracted into local to prevent captures.
+                lazyDiagnostics = GetLazyDiagnostics();
+            }
+
+            var result = await lazyDiagnostics.GetValueAsync(cancellationToken).ConfigureAwait(false);
+            return (ImmutableArray<DiagnosticData>)result;
+
+            AsyncLazy<IReadOnlyList<DiagnosticData>> GetLazyDiagnostics()
+            {
+                return s_projectToDiagnostics.GetValue(
+                    Document.Project,
+                    _ => AsyncLazy.Create<IReadOnlyList<DiagnosticData>>(
+                        async cancellationToken => await diagnosticAnalyzerService.GetDiagnosticsForIdsAsync(
+                            Document.Project.Solution, Document.Project.Id, documentId: null,
+                            diagnosticIds: null, shouldIncludeAnalyzer,
+                            // Ensure we compute and return diagnostics for both the normal docs and the additional docs in this project.
+                            static (project, _) => [.. project.DocumentIds.Concat(project.AdditionalDocumentIds)],
```

Comment:
@mavasani i needed to be able to do this as otherwise the built-in logic of the diag-analyzer-service is just to get the diagnostics for `project.DocumentIds` alone.  I considered tweaking the core logic, but there were a fair amount of callers of `GetDiagnosticsForIdsAsync` and i was concerned that they might break if they started getting diags for additional-docs as well as normal docs.